### ISSUE TITLE
Upgrading from V4.18 (SQL Server 2008) to V5.2 (SQL Server 2016) causes query error: ActiveRecord::StatementInvalid (TinyTds::Error: Incorrect syntax near 'OFFSET'

### ISSUE THREAD

**Issue by patsch** [rails-sqlserver/activerecord-sqlserver-adapter#712]
I am upgrading a rails app from 4.2 to 5.2 ; the Rails upgrade requires me to upgrade the ActiveRecord SQL Server Adapter to 5.2 as well.

After the upgrade I get this error when trying to load a model:
`>> User.find(1)`

==>

`ActiveRecord::StatementInvalid (TinyTds::Error: Incorrect syntax near 'OFFSET'.: EXEC sp_executesql N'SELECT  [Users].* FROM [Users] WHERE [Users].[idUsers] = @0  ORDER BY [Users].[idUsers] ASC OFFSET 0 ROWS FETCH NEXT @1 ROWS ONLY', N'@0 int, @1 int', @0 = 1, @1 = 1)`

This seems unrelated to the underlying database upgrade from SQL Server 2008 to SQL Server 2016. When I run the same query in the Rails 4 app (using the 4.1.8 version of the activerecord-sqlserver-adapter gem) against the SQL Server 2016 database it works fine, reason being that when using the 4.1.8 version the same query is sent through like this:

`EXEC sp_executesql N'SELECT TOP (1) [Users].* FROM [Users] WHERE [Users].[idUsers] = 1 ORDER BY [Users].[idUsers] ASC'`

Microsoft recommends using OFFSET and FETCH rather than TOP, but...it's just not working for me :-(

**Comment by patsch** [comment]
Ok, finally got somewhere on this. This seems to have to do with SQL Server compatibility levels:

http://www.smattie.com/2017/01/30/sql-server-compatibility-levels/

After restoring the database backup from SQL Server 2008 into SQL Server 2016 it is still set to be 'SQL Server 2008 compatible'.

Hence why the OFFSET parameter is NOT recognised:

```
SELECT compatibility_level  
 FROM sys.databases WHERE name = 'xxx';  
``` 

=> 100

```
 65 - SQL Server 6.5
 70 - SQL Server 7.0
 80 - SQL Server 2000
 90 - SQL Server 2005
100 - SQL Server 2008/R2
110 - SQL Server 2012
120 - SQL Server 2014
130 - SQL Server 2016
140 - SQL Server 2017
```

See  https://stackoverflow.com/questions/42263984/incorrect-syntax-near-offset-command

So - apologies. Looks like nothing to do with the Adapter after all, just need to figure out how to make the old backup *compatible*

**Comment by patsch** [comment]
Well, it looked like a good explanation - unfortunately after I changed the compatibility levels to 130...
`ALTER DATABASE XXX SET COMPATIBILITY_LEVEL = 130`

and restarted the server I am still getting the same error as before.

When I manually create a TinyTds client object against the database, it still claims that the version of my server is SQL Server 2008:

`client = TinyTds::Client.new username: SECRET['DB_DEV_USER'], password: SECRET['DB_DEV_PASS'], host: SECRET['DB_DEV_HOST'], database: 'XXX'
 => #<TinyTds::Client:0x00007fa777e08b40 @message_handler=nil, @query_options={:as=>:hash, :symbolize_keys=>false, :cache_rows=>true, :timezone=>:local, :empty_sets=>true}>
2.6.5 :002 > client.tds_version_info
 => "DBTDS_7_3 - Microsoft SQL Server 2008"
2.6.5 :003 >`

Could this be an explanation, i.e. is TinyTDS asking SQL Server - for some reason - to act in the lower compatibility mode ? 

Any help in resolving this would be highly appreciated.

**Comment by patsch** [comment]
When I execute the raw SQL
`EXEC sp_executesql N'SELECT  [Users].* FROM [Users] WHERE [Users].[idUsers] = @0  ORDER BY [Users].[idUsers] ASC OFFSET 0 ROWS FETCH NEXT @1 ROWS ONLY', N'@0 int, @1 int', @0 = 1, @1 = 1`

in the SQL Server Management console, it works fine ! But if the same statement is sent from Rails 5.2 via the ActiveRecord SQL Server Adapter 5.2 via TinyTds 2.1.2 to the Server it fails with the given error: `Incorrect syntax near 'OFFSET'`

**Comment by patsch** [comment]
finally fixed. was ultimately just a config problem on MY end; the app uses 3 databases; I forgot to adjust the configuration for the secondary ones; but at least it lead me to the compatibility setting on the Sql server end, which I would have needed to make this work.

**Comment by jeremysenn** [comment]
I have an issue that seems somewhat similar to this, but a bit more peculiar. We recently upgraded to SQL Server 2019, and now have some problems properly connecting to the database. 

tiny_tds: 2.1.2
activerecord-sqlserver-adapter: 6.0.0

The first issue that we noticed the  was the "Incorrect syntax near 'OFFSET'" when doing any searches of the database in Ruby (e.g. Image.find(1) ).

Upon further investigation we found the following:

`Image.columns_hash => {}`

However, we have no problem listing columns via the console when when using the tds client directly:

`results = client.execute("SELECT * FROM INFORMATION_SCHEMA.COL MNS WHERE TABLE_NAME = N'images'")`

From the client in the console the compatibility level is correct:

`{"compatability_level" => 150} `

However, client.tds_version_info gives:

`"DBTDS_7_3 - Microsoft SQL Server 2008"`

Any help would be greatly appreciated!
